### PR TITLE
Add search range for `search-pattern`

### PR DESCRIPTION
**Pull request by daniellimws**
### Description/Motivation/Screenshots ###
This adds an argument for the command `search-pattern` (`grep`, `xref`) for the user to choose an address range or section to search from.

This is because sometimes one may need to search for a very common pattern, and having the terminal flooded with output is troublesome.

ss

This argument comes before the one for endianness. To set endianness without limiting the search space, `grep 0x400123 memory big`.

### How Has This Been Tested? ###

| Architecture | Yes/No                   | Comments               |
|--------------|:------------------------:|------------------------|
| x86-32       | :heavy_multiplication_x: | :heavy_check_mark: |
| x86-64       | :heavy_multiplication_x: | :heavy_check_mark: |
| ARM          | :heavy_multiplication_x: |                        |
| AARCH64      | :heavy_multiplication_x: |                        |
| MIPS         | :heavy_multiplication_x: |                        |
| POWERPC      | :heavy_multiplication_x: |                        |
| SPARC        | :heavy_multiplication_x: |                        |
| `make tests` | :heavy_multiplication_x: |                        |

### Checklist ###

- [ ] If my code is a bug fix it targets master, otherwise it targets dev.
- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required.
- [ ] My change adds tests as appropriate.
- [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by daniellimws** [comment]
🤦‍♂️ tried to rebase with dev first

**Comment by Grazfather** [comment]
Fix the commit history here, the diff is messed up.

We shouldn't add args to a command before other args -- This messes people up who are used to it working one way. We could add it as an optional arg at the end

**Comment by daniellimws** [comment]
Should be ok now?

**Comment by hugsy** [comment]
Hey @daniellimws ! Is it all ready for testing and merging ?

**Comment by daniellimws** [comment]
Yes it is.

**Comment by hugsy** [comment]
Then I'll test and merge today.

Thanks man!

File: gef.py [hunk 1]
```diff
@@ -5106,9 +5106,21 @@ class SearchPatternCommand(GenericCommand):
...
+    _example_ = "\n{0:s} AAAAAAAA\n{0:s} 0x555555554000 little stack\n{0:s}AAAA 0x600000-0x601000".format(_cmdline_)
+
+    def print_section(self, section):
+        title = "In "
+        if section.path:
+            title += "'{}'".format(Color.blueify(section.path) )
+
+        title += "({:#x}-{:#x})".format(section.page_start, section.page_end)
+        title += ", permission={}".format(section.permission)
+        ok(title)
```
Comment (hugsy) [inline gef.py]: Missing return

File: gef.py [hunk 2]
```diff
@@ -5106,9 +5106,21 @@ class SearchPatternCommand(GenericCommand):
...
+
+    def print_loc(self, loc):
+        gef_print("""  {:#x} - {:#x} {}  "{}" """.format(loc[0], loc[1], RIGHT_ARROW, Color.pinkify(loc[2]),))
```
Comment (hugsy) [inline gef.py]: Ditto

File: gef.py [hunk 3]
```diff
@@ -5179,12 +5180,36 @@ def do_invoke(self, argv):
...
         pattern = argv[0]
         endian = get_endian()
-        if argc==2:
-            if argv[1]=="big": endian = Elf.BIG_ENDIAN
-            elif argv[1]=="small": endian = Elf.LITTLE_ENDIAN
 
-        info("Searching '{:s}' in memory".format(Color.yellowify(pattern)))
-        self.search_pattern(pattern, endian)
+        if argc >= 2:
+            if argv[1] == "big": endian = Elf.BIG_ENDIAN
```
Comment (hugsy) [inline gef.py]: Replace with `argv[1].lower()` just in case. Doesn't cost much and avoid potential Issues

File: gef.py [hunk 4]
```diff
@@ -5179,12 +5180,36 @@ def do_invoke(self, argv):
...
+            elif argv[1] == "small": endian = Elf.LITTLE_ENDIAN
```
Comment (hugsy) [inline gef.py]: Same

File: gef.py [hunk 5]
```diff
@@ -5179,12 +5180,36 @@ def do_invoke(self, argv):
...
+            else:
+                pattern = "".join(["\\x"+pattern[i:i+2] for i in range(len(pattern) - 2, 0, -2)])
+
+        if argc == 3:
+            info("Searching '{:s}' in {:s}".format(Color.yellowify(pattern), argv[2]))
+
+            if "0x" in argv[2]:
+                start, end = parse_string_range(argv[2])
+
+                self.print_section(lookup_address(start).section)
```
Comment (hugsy) [inline gef.py]: `lookup_address()` can return an `Address` marked as invalid (`addr.valid = False`). We should test that before using it in the `print_section()` call.

**Comment by daniellimws** [comment]
Ok it should be good to go now

**Comment by hugsy** [comment]
I am merging this since it's a very small change, and all pure Python so there's no need to test on all arch.

**Pull request merged**